### ISSUE TITLE
explore acceleration w WebAssembly

### ISSUE THREAD

**Issue by jywarren** [publiclab/image-sequencer#400]
WebAssembly seems like a way we might speed up pixel-by-pixel operations. 

https://developer.mozilla.org/en-US/docs/WebAssembly

But would it suffer from the same issues as webworkers, i.e. a cost of passing the image data over?

**Comment by jywarren** [comment]
Also note this way of writing WebAssembly:

https://github.com/ballercat/walt

**Comment by tech4GT** [comment]
@jywarren Even I was thinking we can write the pixel manipulation code in c and create a js binding to it, is that the same as web assembly? I mean the code shows up as native code.

**Comment by jywarren** [comment]
Interesting -- well, the WebAssembly approach means it'd be runnable at (near?) native speed in the browser, too! I think it could get really fast -- maybe not GPU speed, but maybe closer to video speed. For a speed reference we'd like to get closer to, see the Infragram WebGL demo: https://publiclab.github.io/infragram/

**Comment by jywarren** [comment]
Walt seems like a good way to easily write webassembly code! 

> Provide a thin layer of syntax sugar on top of .wat text format. Preferably porting as much of JavaScript syntax to WebAssembly as possible. This improved syntax should give direct control over the WebAssembly output. Meaning there should be minimal to none post optimization to be done to the wast code generated. The re-use of JavaScript semantics is intentional as I do not wish to create a brand new language.

**Comment by jywarren** [comment]
Also noting imagemagick has been compiled for webassembly... Wow!

https://github.com/KnicKnic/WASM-ImageMagick/blob/master/Readme.md

**Comment by jywarren** [comment]
I think we could write a version of Pixelmanipulation following this example!

https://github.com/ballercat/walt/wiki/Walt-In-5-Minutes

...but would we be able to run a JavaScript function inside? I'm not sure about that. Would the function have to be compiled as well?

**Comment by jywarren** [comment]
Docs for reading an image into webassembly: https://developer.mozilla.org/en-US/docs/WebAssembly/existing_C_to_wasm

**Comment by jywarren** [comment]
This is a great area for testing code snippets!

https://ballercat.github.io/walt/

**Comment by jywarren** [comment]
Just asked over at the Walt project if their compiler runs in the browser. Otherwise we could look at precompiling these at build time.

**Comment by jywarren** [comment]
So I think if the simple pixel manipulation function we pass into the PixelManipulation module is close enough to Walt syntax, which is actually WebAssembly, but "looks like" javascript... so if the inside of a function like this could be compiled almost directly, that'd be interesting:

https://github.com/publiclab/image-sequencer/blob/1311119787e6b7e7100cec80da59253adb0eba8b/src/modules/Brightness/Module.js#L22-L29

**Comment by jywarren** [comment]
Walt folks responded to say that `walt` does indeed compile in the browser, on the fly! Awesome.

**Comment by Mridul97** [comment]
@jywarren @tech4GT I was reading the above discussions.. I was thinking what would be a better way, do we write pixel manipulation code in c and then compile it to wasm and then include it in our project or we write pixel manipulation code in walt and it will compile in the browser? Anyways WebAssembly seems to be a great way to increase speed! :smile:

**Comment by jywarren** [comment]
that sounds great!

**Comment by Divy123** [comment]
Can I work on this issue?

**Comment by jywarren** [comment]
yes please!!!

**Comment by HarshKhandeparkar** [comment]
@jywarren @publiclab/is-reviewers [nectarJs](https://codeburst.io/nectarjs-compiling-javascript-into-native-binaries-for-every-platform-2efb2083a4a) claims that it can compile js to native OS binaries for all platforms. This can be used to compile the lib for CLI maybe?

NectarJs [github page](https://github.com/NectarJS/nectarjs) [webpage](https://nectar-lang.com)

**Comment by Divy123** [comment]
@jywarren your thoughts on nectarJs please?

**Comment by jywarren** [comment]
Hmm, i thought webassembly is cross-platform anyways? But that is very cool!

Ah, i see that nectarJs seems to have an option to compile to WASM (which
should then be cross-platform as that's the WebAssembly format). Perhaps we
could do that with some portions of our code, like PixelManipulation, to
achieve our goal?

**Comment by Divy123** [comment]
@jywarren in that case may I proceed with nectarJs for applying the same to our PixelManipulation.js ?

**Comment by jywarren** [comment]
I would love to see a prototype of this in a PR, yes please!

**Comment by Divy123** [comment]
Excited for this.
Thanks!

**Comment by jywarren** [comment]
same!!!

**Comment by jywarren** [comment]
Really nice walkthrough of webassembly here! https://medium.freecodecamp.org/get-started-with-webassembly-using-only-14-lines-of-javascript-b37b6aaca1e4

**Comment by HarshKhandeparkar** [comment]
@jywarren just a note. In the NectarJs repo, there was a discussion about it not being fully open source. I don't know what it meant. I know one thing for sure and that is that the user needs to have an account on https://nectar-lang.com and should have a key to use NectarJs. Just wanted to warn everybody before using this. The dev said he wanted help from everyone to make Nectar fully open source(maybe it is already done?)

cc @Divy124

**Comment by jywarren** [comment]
Ok! There are other ways of making wasm files listed above if we run into
trouble. Thanks!

**Comment by Divy123** [comment]
Thanks @HarshKhandeparkar .

**Comment by Divy123** [comment]
@jywarren I looked into WASM, made these two little projects,
https://github.com/Divy123/canvas-wasmvas
https://github.com/Divy123/prime-with-wasm

I suggest we can proceed with writing native C code and compiling that with emscripten and utilizing the genrated glue JS files  in our code which I feel can be quite handy.

**Comment by jywarren** [comment]
Ok, however did you see the library to generate wasm from JavaScript like
syntax? In either case, we should make a src/webassembly folder perhaps to
contain our source code and put the wasm files in dist, right? Thanks!!!

**Comment by Divy123** [comment]
Ya I looked into but perhaps feel this can be more easy to go with as then we don't have the overhead of a library.
Yes we have to keep wasm files in src/webassembly and keep it in dist!!
What do you think?

**Comment by jywarren** [comment]
OK, sounds good. We can always refactor to use a different wasm generator
later, so this could even be solved in parts. Best first get a
demonstration running of it working at all!

**Comment by Divy123** [comment]
Sure !!

**Comment by Divy123** [comment]
@jywarren one thing that if we want to convert PixelManipulation.js files and the pixel processing in modules as well.
Do we have something other as well ?
I mean can you please explain a bit more on the applicable areas of our code base here where we want to use wasm

**Comment by Divy123** [comment]
@jywarren just one thing that if you can suggest apart from PixelManipulation.js what are some other areas that we can apply wasm to include in our code base like can we add to different modules as well?
Thanks!!

**Comment by jywarren** [comment]
Hi! Well, my thought was that PixelManipulation is used a LOT so if we
improve that, we'll see improvements across a lot of our codebase. And we
can start there to get a big improvement in one step, and then follow up to
see where else we could apply this. Many modules use PixelManipulation in
the same way, so it's a great leverage point to make a wide-ranging
optimization.

**Comment by tech4GT** [comment]
Yes, this is absolutely true, the heavy lifting is done by pixelManipulation for almost all modules, so that should give us a significant boost in performance!

**Comment by HarshKhandeparkar** [comment]
What abt gpu.js which I am trying to implement for PixelManipulation? It will also work in node and will process the pixels on gpu. Can that also be compiled to wasm? It uses a node library headless-gl(will use in the future). Currently it uses the canvas api.

**Comment by Divy123** [comment]
Hi Harsh!
Can you explain a bit of your work here with gpu.js?

**Comment by HarshKhandeparkar** [comment]
You can have a look at issue #849. Essentially it will manipulate all the pixels at the same time in parallel on a gpu.

**Comment by jywarren** [comment]
I think we can pursue gl acceleration in a parallel track and use our new
benchmarks to compare!

**Comment by Divy123** [comment]
@jywarren @tech4GT I have been through PixelManipulation.js file and what I am able to understand that the whole code here can not be converted as there are lots of things like get-pixels and save-pixels that have to be passed to our C code. So can I get some help regarding exactly what parts of the code do we want to convert to C code or the whole PixelManipulation.js code?

Thanks !!

**Comment by Divy123** [comment]
@jywarren can you please help on this?

**Comment by jywarren** [comment]
Also noting @MargaretAN9 is interested in this! 

Hmm. I think this may be a good question to ask @HarshKhandeparkar for a bit more info here. I had hoped we could take everything /between/ get-pixels and save-pixels and run that in webassembly.

**Comment by Divy123** [comment]
Ok sure!! Thanks!

@MargaretAN9 can we collaborate on this??
Also this is my this week's target.
@HarshKhandeparkar any insights?

@jywarren this is my task for this week. SO should I start working on this or wait for Harsh's reply?

**Comment by jywarren** [comment]
If we don't get input from others in time, perhaps we could swap this
week's task with another to give people a little more time? Thanks!

**Comment by Divy123** [comment]
@jywarren can I just send here a PR with a working demo and then proceed with everyone's suggestions. Meanwhile I can make preparations for the next weeks as well to speed up the things there??

**Comment by jywarren** [comment]
oh that'd be great! always great to see working code to offer input on.
thank you!!!

**Comment by Divy123** [comment]
Thanks a lot Jeff!! :smile:

**Comment by Divy123** [comment]
![Screenshot from 2019-05-30 01-26-27](https://user-images.githubusercontent.com/31225007/58587048-0bf90180-827a-11e9-874e-280e9fc6fb86.png)
I have made a wasm file from wasm fiddle and here is my C code 

```
int getPixels(int , int , int );

int setPixels(int, int, int, int);

int * changePixel(int, int, int, int, int, int);

void paceOp();

void consoleLog(int);

void manipulatePixel(int width, int height, int inBrowser, int test){
  consoleLog(100);
  for (int x = 0; x < width; x++) {
        for (int y = 0; y <height; y++) {
          int* pixel =changePixel(
            getPixels(x, y, 0),
            getPixels(x, y, 1),
             getPixels(x, y, 2),
             getPixels(x, y, 3),
            x,
            y
          );
          setPixels(x, y, 0, pixel[0]);
          setPixels(x, y, 1, pixel[1]);
          setPixels(x, y, 2, pixel[2]);
          setPixels(x, y, 3, pixel[3]);

          if (!inBrowser && !test) paceOp();
        }
      }
}
```

I was trying to convert this part of code for testing from PixelManipulation.js file

```

for (var x = 0; x < pixels.shape[0]; x++) {
      for (var y = 0; y < pixels.shape[1]; y++) {
       let pixel = options.changePixel(
         pixels.get(x, y, 0),
         pixels.get(x, y, 1),
          pixels.get(x, y, 2),
             pixels.get(x, y, 3),
           x,
           y
       );

        pixels.set(x, y, 0, pixel[0]);
        pixels.set(x, y, 1, pixel[1]);
          pixels.set(x, y, 2, pixel[2]);
          pixels.set(x, y, 3, pixel[3]);

          if (!options.inBrowser && !process.env.TEST) pace.op();
         }
       }

```

After adding the wasm file, the code looks like:

```
const imports = {
         env:{
           changePixel:options.changePixel,
           getPixels: pixels.get,
           setPixels: pixels.set,
           paceOp: require('pace').op,
           consoleLog:console.log
         }
       }
      const inBrowser = (options.inBrowser)?1:0;
      const test = (process.env.TEST)?1:0
       WebAssembly.instantiateStreaming(fetch('./program.wasm'), imports)
           .then(wasm => {
      wasm.instance.exports.manipulatePixel(pixels.shape[0],pixels.shape[1],inBrowser, test)
           });

```

But I think the wasm file is not getting recognized, is it some bundler issue?

@jywarren @HarshKhandeparkar need your help here!!
Its something urgent. Please help !

**Comment by jywarren** [comment]
Hi @Divy123 -- can you share a PR link, so we can take a look at how you've gotten it set up? Help us reproduce your error so we can help you solve it! I've also pinged folks in the weekly check-in to see if anyone there can help.

**Comment by Divy123** [comment]
Thanks a lot @jywarren !!
Making a PR now

**Comment by Divy123** [comment]
@jywarren here is the link:
https://github.com/publiclab/image-sequencer/pull/1093

Please look into it.
